### ISSUE TITLE
[Accessibility] Support speech typing in terminal

### ISSUE THREAD

**Issue by jooyoungseo** [microsoft/vscode#199128]
Type: <b>Feature Request</b>

CC @meganrogge 

`ms-vscode.vscode-speech` currently supports Copilot Chat only. However, its speech-input feature can empower people with dis/abilities to improve their typing performance in other editor areas.

I understand that the speech-typing is primarily trained for natural languages rather than programming syntax; however, the natural speech typing in editor areas is still useful for  writing comments, drafting markdown/plaintext documents, etc.

In terminal, the natural language speech input can be used in conjunction with `github-copilot-cli` tool so there is no reason to prevent users from voice-typing through this extension in terminal.

Extension version: 0.3.2023110115
VS Code version: Code - Insiders 1.85.0-insider (eb56719ae7653bff54989385818ef39e1cd73d29, 2023-11-24T15:21:39.004Z)
OS version: Windows_NT x64 10.0.22631
Modes:

**Comment by bpasero** [comment]
@meganrogge @Tyriar this is all ready to be adopted in the workbench, simply use `ISpeechService.createSpeechToTextSession` and listen to the `onDidChange` events. You can use `ISpeechService.hasSpeechProvider` to know if there is an extension present.

**Comment by meganrogge** [comment]
Natural language search in the quick pick / command palette would also be nice

**Comment by bpasero** [comment]
I think eventually we would want to support speech input to anywhere you can type (input fields, text areas).

**Comment by meganrogge** [comment]
One question is how we handle the insertion of special characters /punctuation via voice input. 

@jooyoungseo mentioned that some software allows mapping between a nonverbal sound and punctuation (like a tongue click)

**Comment by meganrogge** [comment]
@bpasero for these, what do you think about a status bar indicator when the speech to text session is in progress for sighted users?

tracking the issue for screen reader users here https://github.com/microsoft/vscode/issues/201901

**Comment by meganrogge** [comment]
@jooyoungseo how do you prevent VS code's speech detection from taking in the screen reader content? When I enable voiceover and try to use VS Code speech, parts of what the screen reader says are inputted.

**Comment by bpasero** [comment]
@meganrogge yes I would think a global indicator makes sense, but maybe it needs to account for indicators that are already in context: for example, the copilot chat input box indicates recording. I would not want to see more indicators globally, e.g. in the status bar.

That makes me wonder if maybe all indication should be in context, for example where the cursor is in the editor or terminal.

//cc @hbons as I think this is also a UX issue to look at

**Comment by meganrogge** [comment]
We could perhaps put an indicator on the editor / terminal tab

**Comment by bpasero** [comment]
I have pinged you on a UX design from Jo that I found interesting for inline chat.

**Comment by meganrogge** [comment]
editor inspiration:

<img width="513" alt="Screenshot 2024-01-11 at 7 16 42 AM" src="https://github.com/microsoft/vscode/assets/29464607/37b9bd7f-90da-4e45-83ec-25c1b580fafa">

**Comment by meganrogge** [comment]
Going to track speech typing in the editor / input fields in a separate issue
